<b><font size="5">Segmenting and Clustering of Toronto Neighborhood</font></b>

<b>1. Import Libraries </b>

In [1]:
#import libraries
import pandas as pd
import numpy as np

# tranform JSON file into a pandas dataframe
from pandas.io.json import json_normalize 

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

#import folium to render map
!conda install -c conda-forge folium=0.5.0 --yes
import folium

print("Imported!")

Fetching package metadata .............
Solving package specifications: .

Package plan for installation in environment /opt/conda/envs/DSX-Python35:

The following NEW packages will be INSTALLED:

    altair:  2.2.2-py35_1 conda-forge
    branca:  0.3.1-py_0   conda-forge
    folium:  0.5.0-py_0   conda-forge
    vincent: 0.4.4-py_1   conda-forge

altair-2.2.2-p 100% |################################| Time: 0:00:00  54.10 MB/s
branca-0.3.1-p 100% |################################| Time: 0:00:00  35.12 MB/s
vincent-0.4.4- 100% |################################| Time: 0:00:00  39.68 MB/s
folium-0.5.0-p 100% |################################| Time: 0:00:00  47.67 MB/s
Imported!


In [2]:
#Install beautifulsoup and html parser
!easy_install beautifulsoup4
!easy_install html5lib

Searching for beautifulsoup4
Best match: beautifulsoup4 4.6.0
Adding beautifulsoup4 4.6.0 to easy-install.pth file

Using /opt/conda/envs/DSX-Python35/lib/python3.5/site-packages
Processing dependencies for beautifulsoup4
Finished processing dependencies for beautifulsoup4
Searching for html5lib
Best match: html5lib 0.999999999
Adding html5lib 0.999999999 to easy-install.pth file

Using /opt/conda/envs/DSX-Python35/lib/python3.5/site-packages
Processing dependencies for html5lib
Finished processing dependencies for html5lib


<b>2. Get the web page contents from the Wikipedia web site</b>

In [3]:
#Get page contents from the given URL
import requests
page = requests.get('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M')
#print(page.content)


<b>3. Use the HTML parser to extract table from the web page</b>

In [4]:
#Using BeautifulSoup to extract the page
from bs4 import BeautifulSoup
soup = BeautifulSoup(page.content, 'html.parser')

<p>
First, get contents in tags 'tr' from the table.
Next, extract the contents within the header tags 'th'
Last, extract the data within tags 'td' if Borough is not 'Not assigned'.
Change the value of Neighbourhood if its value is 'Not assigned'.
Save the data to a dataframe.
</p>

In [5]:
#Extract table from the page
html = list(soup.children)[2]
list(html.children)
body = list(html.children)[3]
body
canada_table = soup.find_all('table')[0]

list_tr = canada_table.find_all("tr")

#Extract header tags
header = []
tr = list_tr[0]
for th in tr:
    if(th != "\n"):
        header.append(th.get_text().rstrip())

#Read table contents and change Borough to Postcode if Borough is 'Not assigned'
records = []
for tr in list_tr:
    trs = tr.findAll("td")
    if trs:
        record = []
        record.append(trs[0].text)
        #Check if Borough is 'Not assigned'
        if(trs[1].text != "Not assigned"):
            record.append(trs[1].text)
            #Check if 'Neighbourhood if 'Not assigned'
            if(trs[2].text.rstrip() == "Not assigned"):
                record.append(trs[1].text)
            else:
                record.append(trs[2].text.rstrip())
            records.append(record)
    
#Create dataframe with table contents
canada_df = pd.DataFrame(data=records, columns = header)
print(canada_df.shape)
#print(canada_df)


(212, 3)


In [6]:
#Find rows with same Postcode and Borough and concat the Neighbourhood names 
#Create a new dataframe with these changes
list_canada = []
for (k1, k2), group in canada_df.groupby(['Postcode', 'Borough']):
    col = group['Neighbourhood']
    names = pd.Series(group['Neighbourhood'].values)
    grouped_names = names.str.cat(sep=',')
    new_row = [k1, k2, grouped_names]
    list_canada.append(new_row)

new_canada_df=pd.DataFrame(list_canada,columns=['Postcode','Borough', 'Neighbourhood'])
print(new_canada_df.shape)
print(new_canada_df.head(40))

(103, 3)
   Postcode       Borough                                      Neighbourhood
0       M1B   Scarborough                                      Rouge,Malvern
1       M1C   Scarborough               Highland Creek,Rouge Hill,Port Union
2       M1E   Scarborough                    Guildwood,Morningside,West Hill
3       M1G   Scarborough                                             Woburn
4       M1H   Scarborough                                          Cedarbrae
5       M1J   Scarborough                                Scarborough Village
6       M1K   Scarborough          East Birchmount Park,Ionview,Kennedy Park
7       M1L   Scarborough                      Clairlea,Golden Mile,Oakridge
8       M1M   Scarborough      Cliffcrest,Cliffside,Scarborough Village West
9       M1N   Scarborough                         Birch Cliff,Cliffside West
10      M1P   Scarborough  Dorset Park,Scarborough Town Centre,Wexford He...
11      M1R   Scarborough                                   Maryval

<b>4. Read CSV file with Geospatial data of Canada to a dataframe and merge with the Canada data</b>

In [7]:
#Read the csv file that has Latitude and Longitude data to a DataFrame
lat_lon_data = pd.read_csv('http://cocl.us/Geospatial_data')
lat_lon_data.columns = ['Postcode', 'Latitude', 'Longitude']
lat_lon_data.head(5)


Postcode   Latitude  Longitude
0      M1B  43.806686 -79.194353
1      M1C  43.784535 -79.160497
2      M1E  43.763573 -79.188711
3      M1G  43.770992 -79.216917
4      M1H  43.773136 -79.239476

In [9]:
#Merge canada dataframe with the Latitude and Longitude data
all_canada_df = pd.merge(new_canada_df, lat_lon_data, on = ['Postcode'])
all_canada_df

Postcode           Borough  \
0        M1B       Scarborough   
1        M1C       Scarborough   
2        M1E       Scarborough   
3        M1G       Scarborough   
4        M1H       Scarborough   
5        M1J       Scarborough   
6        M1K       Scarborough   
7        M1L       Scarborough   
8        M1M       Scarborough   
9        M1N       Scarborough   
10       M1P       Scarborough   
11       M1R       Scarborough   
12       M1S       Scarborough   
13       M1T       Scarborough   
14       M1V       Scarborough   
15       M1W       Scarborough   
16       M1X       Scarborough   
17       M2H        North York   
18       M2J        North York   
19       M2K        North York   
20       M2L        North York   
21       M2M        North York   
22       M2N        North York   
23       M2P        North York   
24       M2R        North York   
25       M3A        North York   
26       M3B        North York   
27       M3C        North York   
28       M3H        North York   
29       M3J        North York   
..       ...               ...   
73       M6C              York   
74       M6E              York   
75       M6G  Downtown Toronto   
76       M6H      West Toronto   
77       M6J      West Toronto   
78       M6K      West Toronto   
79       M6L        North York   
80       M6M              York   
81       M6N              York   
82       M6P      West Toronto   
83       M6R      West Toronto   
84       M6S      West Toronto   
85       M7A      Queen's Park   
86       M7R       Mississauga   
87       M7Y      East Toronto   
88       M8V         Etobicoke   
89       M8W         Etobicoke   
90       M8X         Etobicoke   
91       M8Y         Etobicoke   
92       M8Z         Etobicoke   
93       M9A         Etobicoke   
94       M9B         Etobicoke   
95       M9C         Etobicoke   
96       M9L        North York   
97       M9M        North York   
98       M9N              York   
99       M9P         Etobicoke   
100      M9R         Etobicoke   
101      M9V         Etobicoke   
102      M9W         Etobicoke   

                                         Neighbourhood   Latitude  Longitude  
0                                        Rouge,Malvern  43.806686 -79.194353  
1                 Highland Creek,Rouge Hill,Port Union  43.784535 -79.160497  
2                      Guildwood,Morningside,West Hill  43.763573 -79.188711  
3                                               Woburn  43.770992 -79.216917  
4                                            Cedarbrae  43.773136 -79.239476  
5                                  Scarborough Village  43.744734 -79.239476  
6            East Birchmount Park,Ionview,Kennedy Park  43.727929 -79.262029  
7                        Clairlea,Golden Mile,Oakridge  43.711112 -79.284577  
8        Cliffcrest,Cliffside,Scarborough Village West  43.716316 -79.239476  
9                           Birch Cliff,Cliffside West  43.692657 -79.264848  
10   Dorset Park,Scarborough Town Centre,Wexford He...  43.757410 -79.273304  
11                                    Maryvale,Wexford  43.750072 -79.295849  
12                                           Agincourt  43.794200 -79.262029  
13               Clarks Corners,Sullivan,Tam O'Shanter  43.781638 -79.304302  
14   Agincourt North,L'Amoreaux East,Milliken,Steel...  43.815252 -79.284577  
15                        L'Amoreaux West,Steeles West  43.799525 -79.318389  
16                                         Upper Rouge  43.836125 -79.205636  
17                                   Hillcrest Village  43.803762 -79.363452  
18                          Fairview,Henry Farm,Oriole  43.778517 -79.346556  
19                                     Bayview Village  43.786947 -79.385975  
20                             Silver Hills,York Mills  43.757490 -79.374714  
21                              Newtonbrook,Willowdale  43.789053 -79.408493  
22                                    Willowdale South  43.770120 -79.408493  

<b>5. Extracting only Toronto data from the Canada data</b>

In [16]:
#Create a dataframe with rows that have Toronto in Borough
list_toronto = []
toronto_df = pd.DataFrame(columns = ['Postcode','Borough', 'Neighbourhood', 'Latitude', 'Longitude'])

for (k1), group in all_canada_df.groupby(['Borough']):
    if 'Toronto' in k1:
        toronto_df = toronto_df.append(group)

toronto_df.reset_index(drop=True, inplace=True)

print(toronto_df.shape)
print(toronto_df.head(38))

                

(38, 5)
   Postcode           Borough  \
0       M4N   Central Toronto   
1       M4P   Central Toronto   
2       M4R   Central Toronto   
3       M4S   Central Toronto   
4       M4T   Central Toronto   
5       M4V   Central Toronto   
6       M5N   Central Toronto   
7       M5P   Central Toronto   
8       M5R   Central Toronto   
9       M4W  Downtown Toronto   
10      M4X  Downtown Toronto   
11      M4Y  Downtown Toronto   
12      M5A  Downtown Toronto   
13      M5B  Downtown Toronto   
14      M5C  Downtown Toronto   
15      M5E  Downtown Toronto   
16      M5G  Downtown Toronto   
17      M5H  Downtown Toronto   
18      M5J  Downtown Toronto   
19      M5K  Downtown Toronto   
20      M5L  Downtown Toronto   
21      M5S  Downtown Toronto   
22      M5T  Downtown Toronto   
23      M5V  Downtown Toronto   
24      M5W  Downtown Toronto   
25      M5X  Downtown Toronto   
26      M6G  Downtown Toronto   
27      M4E      East Toronto   
28      M4K      East Toronto   
29

In [13]:
print('The dataframe has {} boroughs and {} neighborhoods.'.format(
        len(toronto_df['Borough'].unique()),
        toronto_df.shape[0]
    )
)

The dataframe has 4 boroughs and 38 neighborhoods.


<b>6. Get Geographical coordinates of Toronto and render a map with Neighbourhoods</b>

In [17]:
from geopy.geocoders import Nominatim
address = 'Toronto, Canada'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.653963, -79.387207.


In [18]:

# create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighbourhood in zip(toronto_df['Latitude'], toronto_df['Longitude'], toronto_df['Borough'], toronto_df['Neighbourhood']):
    label = '{}, {}'.format(neighbourhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

<p><b font size=3>7. Explore Toronto Neighbourhoods</font></b></p>
<p>Next, we are going to start utilizing the Foursquare API to explore the neighborhoods and segment them.</p>

In [61]:
# The code was removed by Watson Studio for sharing.

Initialized Foursquare credentials: client_id, client_secret, and version in this cell.


In [20]:
toronto_df.loc[0, 'Neighbourhood']

'Lawrence Park'

Get the neighbourhood's Latitude and Longitude

In [21]:
neighbourhood_latitude = toronto_df.loc[0, 'Latitude'] # neighbourhood latitude value
neighbourhood_longitude = toronto_df.loc[0, 'Longitude'] # neighbourhood longitude value

neighbourhood_name = toronto_df.loc[0, 'Neighbourhood'] # neighbourhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighbourhood_name, 
                                                               neighbourhood_latitude, 
                                                               neighbourhood_longitude))

Latitude and longitude values of Lawrence Park are 43.7280205, -79.3887901.


<B>Geting top 100 venues that are in Lawrence Park within a radius of 500 meters.<B>

In [63]:
search_query = 'Lawrence Park'
radius = 500
LIMIT = 100
url = 'https://api.foursquare.com/v2/venues/search?client_id={}&client_secret={}&ll={},{}&v={}&query={}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    neighbourhood_latitude, 
    neighbourhood_longitude, 
    VERSION, 
    search_query, 
    radius, LIMIT)
print("Created url with Foursquare credentials:'https://api.foursquare.com/v2/venues/search?client_id=*******&client_secret=************&ll=43.7280205,-79.3887901&v=20180605&query=Lawrence Park&radius=500&limit=100'")

Created url with Foursquare credentials:'https://api.foursquare.com/v2/venues/search?client_id=*******&client_secret=************&ll=43.7280205,-79.3887901&v=20180605&query=Lawrence Park&radius=500&limit=100'


In [25]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5c5e0779dd5797192ad688e7'},
 'response': {'venues': [{'categories': [{'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/parks_outdoors/park_',
       'suffix': '.png'},
      'id': '4bf58dd8d48988d163941735',
      'name': 'Park',
      'pluralName': 'Parks',
      'primary': True,
      'shortName': 'Park'}],
    'hasPerk': False,
    'id': '50e6da19e4b0d8a78a0e9794',
    'location': {'address': '3055 Yonge Street',
     'cc': 'CA',
     'city': 'Toronto',
     'country': 'Canada',
     'crossStreet': 'Lawrence Avenue East',
     'distance': 465,
     'formattedAddress': ['3055 Yonge Street (Lawrence Avenue East)',
      'Toronto ON',
      'Canada'],
     'labeledLatLngs': [{'label': 'display',
       'lat': 43.72696303913755,
       'lng': -79.39438246708775}],
     'lat': 43.72696303913755,
     'lng': -79.39438246708775,
     'state': 'ON'},
    'name': 'Lawrence Park Ravine',
    'referralId': 'v-1549666169'},
   {'categories': [{'ico

In [26]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [33]:

venues = results['response']['venues'] 
nearby_venues = json_normalize(venues) # flatten JSON

filtered_columns = ['name', 'categories', 'location.lat', 'location.lng']
nearby_venues = nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['categories'] = nearby_venues.apply(get_category_type, axis=1)
# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues

name    categories        lat        lng
0            Lawrence Park Ravine          Park  43.726963 -79.394382
1  TTC Bus #162 - Lawrence-Donway      Bus Line  43.728026 -79.382805
2                    Wanless Park  Neighborhood  43.727064 -79.394406

In [29]:
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))

3 venues were returned by Foursquare.


<b>Explore all neighbourhoods in Toronto</b>

In [34]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [31]:
toronto_venues = getNearbyVenues(names=toronto_df['Neighbourhood'],
                                   latitudes=toronto_df['Latitude'],
                                   longitudes=toronto_df['Longitude']
                                  )

Lawrence Park
Davisville North
North Toronto West
Davisville
Moore Park,Summerhill East
Deer Park,Forest Hill SE,Rathnelly,South Hill,Summerhill West
Roselawn
Forest Hill North,Forest Hill West
The Annex,North Midtown,Yorkville
Rosedale
Cabbagetown,St. James Town
Church and Wellesley
Harbourfront,Regent Park
Ryerson,Garden District
St. James Town
Berczy Park
Central Bay Street
Adelaide,King,Richmond
Harbourfront East,Toronto Islands,Union Station
Design Exchange,Toronto Dominion Centre
Commerce Court,Victoria Hotel
Harbord,University of Toronto
Chinatown,Grange Park,Kensington Market
CN Tower,Bathurst Quay,Island airport,Harbourfront West,King and Spadina,Railway Lands,South Niagara
Stn A PO Boxes 25 The Esplanade
First Canadian Place,Underground city
Christie
The Beaches
The Danforth West,Riverdale
The Beaches West,India Bazaar
Studio District
Business Reply Mail Processing Centre 969 Eastern
Dovercourt Village,Dufferin
Little Portugal,Trinity
Brockton,Exhibition Place,Parkdale Villag

In [35]:
print(toronto_venues.shape)
toronto_venues.head()

(1694, 7)


Neighbourhood  Neighbourhood Latitude  Neighbourhood Longitude  \
0     Lawrence Park               43.728020               -79.388790   
1     Lawrence Park               43.728020               -79.388790   
2     Lawrence Park               43.728020               -79.388790   
3     Lawrence Park               43.728020               -79.388790   
4  Davisville North               43.712751               -79.390197   

                            Venue  Venue Latitude  Venue Longitude  \
0            Lawrence Park Ravine       43.726963       -79.394382   
1                  Dim Sum Deluxe       43.726953       -79.394260   
2              Zodiac Swim School       43.728532       -79.382860   
3  TTC Bus #162 - Lawrence-Donway       43.728026       -79.382805   
4                   Sherwood Park       43.716551       -79.387776   

       Venue Category  
0                Park  
1  Dim Sum Restaurant  
2         Swim School  
3            Bus Line  
4                Park

<b>Check how many venues are returned for each neighbourhood</b>

In [36]:
toronto_venues.groupby('Neighbourhood').count()

Neighbourhood Latitude  \
Neighbourhood                                                                
Adelaide,King,Richmond                                                 100   
Berczy Park                                                             55   
Brockton,Exhibition Place,Parkdale Village                              19   
Business Reply Mail Processing Centre 969 Eastern                       19   
CN Tower,Bathurst Quay,Island airport,Harbourfr...                      14   
Cabbagetown,St. James Town                                              47   
Central Bay Street                                                      83   
Chinatown,Grange Park,Kensington Market                                100   
Christie                                                                15   
Church and Wellesley                                                    86   
Commerce Court,Victoria Hotel                                          100   
Davisville                                                              37   
Davisville North                                                         9   
Deer Park,Forest Hill SE,Rathnelly,South Hill,S...                      14   
Design Exchange,Toronto Dominion Centre                                100   
Dovercourt Village,Dufferin                                             18   
First Canadian Place,Underground city                                  100   
Forest Hill North,Forest Hill West                                       4   
Harbord,University of Toronto                                           34   
Harbourfront East,Toronto Islands,Union Station                        100   
Harbourfront,Regent Park                                                49   
High Park,The Junction South                                            23   
Lawrence Park                                                            4   
Little Portugal,Trinity                                                 65   
Moore Park,Summerhill East                                               2   
North Toronto West                                                      18   
Parkdale,Roncesvalles                                                   16   
Rosedale                                                                 4   
Roselawn                                                                 1   
Runnymede,Swansea                                                       36   
Ryerson,Garden District                                                100   
St. James Town                                                         100   
Stn A PO Boxes 25 The Esplanade                                         94   
Studio District                                                         40   
The Annex,North Midtown,Yorkville                                       24   
The Beaches                                                              3   
The Beaches West,India Bazaar                                           19   
The Danforth West,Riverdale                                             42   

                                                    Neighbourhood Longitude  \
Neighbourhood                                                                 
Adelaide,King,Richmond                                                  100   
Berczy Park                                                              55   
Brockton,Exhibition Place,Parkdale Village                               19   
Business Reply Mail Processing Centre 969 Eastern                        19   
CN Tower,Bathurst Quay,Island airport,Harbourfr...                       14   
Cabbagetown,St. James Town                                               47   
Central Bay Street                                                       83   
Chinatown,Grange Park,Kensington Market                                 100   
Christie                                                                 15   
Church and Wellesley                                          

In [37]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 235 uniques categories.


<B>Analyze each Neighbourhood in Toronto</B>

In [38]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighbourhood'] = toronto_venues['Neighbourhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Neighbourhood  Adult Boutique  Afghan Restaurant  Airport  \
0     Lawrence Park               0                  0        0   
1     Lawrence Park               0                  0        0   
2     Lawrence Park               0                  0        0   
3     Lawrence Park               0                  0        0   
4  Davisville North               0                  0        0   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                   0             0               0                0   
1                   0             0               0                0   
2                   0             0               0                0   
3                   0             0               0                0   
4                   0             0               0                0   

   Airport Terminal  American Restaurant     ...       Trail  Train Station  \
0                 0                    0     ...           0              0   
1                 0                    0     ...           0              0   
2                 0                    0     ...           0              0   
3                 0                    0     ...           0              0   
4                 0                    0     ...           0              0   

   Vegetarian / Vegan Restaurant  Video Game Store  Video Store  \
0                              0                 0            0   
1                              0                 0            0   
2                              0                 0            0   
3                              0                 0            0   
4                              0                 0            0   

   Vietnamese Restaurant  Wine Bar  Wings Joint  Women's Store  Yoga Studio  
0                      0         0            0              0            0  
1                      0         0            0              0            0  
2                      0         0            0              0            0  
3                      0         0            0              0            0  
4                      0         0            0              0            0  

[5 rows x 236 columns]

In [39]:
toronto_onehot.shape

(1694, 236)

<b>Find mean of occurence of each category for each Neighbourhood in Toronto.</b>

In [40]:
toronto_grouped = toronto_onehot.groupby('Neighbourhood').mean().reset_index()
toronto_grouped

Neighbourhood  Adult Boutique  \
0                              Adelaide,King,Richmond        0.000000   
1                                         Berczy Park        0.000000   
2          Brockton,Exhibition Place,Parkdale Village        0.000000   
3   Business Reply Mail Processing Centre 969 Eastern        0.000000   
4   CN Tower,Bathurst Quay,Island airport,Harbourf...        0.000000   
5                          Cabbagetown,St. James Town        0.000000   
6                                  Central Bay Street        0.000000   
7             Chinatown,Grange Park,Kensington Market        0.000000   
8                                            Christie        0.000000   
9                                Church and Wellesley        0.011628   
10                      Commerce Court,Victoria Hotel        0.000000   
11                                         Davisville        0.000000   
12                                   Davisville North        0.000000   
13  Deer Park,Forest Hill SE,Rathnelly,South Hill,...        0.000000   
14            Design Exchange,Toronto Dominion Centre        0.000000   
15                        Dovercourt Village,Dufferin        0.000000   
16              First Canadian Place,Underground city        0.000000   
17                 Forest Hill North,Forest Hill West        0.000000   
18                      Harbord,University of Toronto        0.000000   
19    Harbourfront East,Toronto Islands,Union Station        0.000000   
20                           Harbourfront,Regent Park        0.000000   
21                       High Park,The Junction South        0.000000   
22                                      Lawrence Park        0.000000   
23                            Little Portugal,Trinity        0.000000   
24                         Moore Park,Summerhill East        0.000000   
25                                 North Toronto West        0.000000   
26                              Parkdale,Roncesvalles        0.000000   
27                                           Rosedale        0.000000   
28                                           Roselawn        0.000000   
29                                  Runnymede,Swansea        0.000000   
30                            Ryerson,Garden District        0.000000   
31                                     St. James Town        0.000000   
32                    Stn A PO Boxes 25 The Esplanade        0.000000   
33                                    Studio District        0.000000   
34                  The Annex,North Midtown,Yorkville        0.000000   
35                                        The Beaches        0.000000   
36                      The Beaches West,India Bazaar        0.000000   
37                        The Danforth West,Riverdale        0.000000   

    Afghan Restaurant   Airport  Airport Food Court  Airport Gate  \
0            0.000000  0.000000            0.000000      0.000000   
1            0.000000  0.000000            0.000000      0.000000   
2            0.000000  0.000000            0.000000      0.000000   
3            0.000000  0.000000            0.000000      0.000000   
4            0.000000  0.071429            0.071429      0.071429   
5            0.000000  0.000000            0.000000      0.000000   
6            0.000000  0.000000            0.000000      0.000000   
7            0.000000  0.000000            0.000000      0.000000   
8            0.000000  0.000000            0.000000      0.000000   
9            0.011628  0.000000            0.000000      0.000000   
10           0.000000  0.000000            0.000000      0.000000   
11           0.000000  0.000000            0.000000      0.000000   
12           0.000000  0.000000            0.000000      0.000000   
13           0.000000  0.000000            0.000000      0.000000   
14           0.000000  0.000000            0.000000      0.000000   
15           0.000000  0.000000            0.000000      0.000000   
16           0.0000

<b>Print each Neighbourhood with top five venues.</b>

In [41]:
num_top_venues = 5

for hood in toronto_grouped['Neighbourhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighbourhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Adelaide,King,Richmond----
                 venue  freq
0          Coffee Shop  0.06
1                 Café  0.05
2  American Restaurant  0.04
3      Thai Restaurant  0.04
4           Steakhouse  0.04


----Berczy Park----
            venue  freq
0     Coffee Shop  0.07
1      Restaurant  0.05
2    Cocktail Bar  0.05
3  Farmers Market  0.04
4     Cheese Shop  0.04


----Brockton,Exhibition Place,Parkdale Village----
            venue  freq
0  Breakfast Spot  0.11
1     Coffee Shop  0.11
2            Café  0.11
3             Gym  0.05
4             Bar  0.05


----Business Reply Mail Processing Centre 969 Eastern----
                venue  freq
0  Light Rail Station  0.11
1          Restaurant  0.05
2          Smoke Shop  0.05
3         Pizza Place  0.05
4                 Spa  0.05


----CN Tower,Bathurst Quay,Island airport,Harbourfront West,King and Spadina,Railway Lands,South Niagara----
              venue  freq
0  Airport Terminal  0.14
1    Airport Lounge  0.14
2   Airport Ser

In [42]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

<b>Display top 10 most common venues of each Neighbourhood.</b>

In [43]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighbourhoods_venues_sorted = pd.DataFrame(columns=columns)
neighbourhoods_venues_sorted['Neighbourhood'] = toronto_grouped['Neighbourhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighbourhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighbourhoods_venues_sorted.head()

Neighbourhood 1st Most Common Venue  \
0                             Adelaide,King,Richmond           Coffee Shop   
1                                        Berczy Park           Coffee Shop   
2         Brockton,Exhibition Place,Parkdale Village        Breakfast Spot   
3  Business Reply Mail Processing Centre 969 Eastern    Light Rail Station   
4  CN Tower,Bathurst Quay,Island airport,Harbourf...        Airport Lounge   

  2nd Most Common Venue 3rd Most Common Venue  4th Most Common Venue  \
0                  Café            Steakhouse    American Restaurant   
1            Restaurant          Cocktail Bar                   Café   
2           Coffee Shop                  Café  Performing Arts Venue   
3           Yoga Studio      Recording Studio             Smoke Shop   
4       Airport Service      Airport Terminal                  Plane   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0       Thai Restaurant        Clothing Store      Asian Restaurant   
1              Beer Bar    Italian Restaurant                Bakery   
2         Burrito Place               Stadium                   Bar   
3            Skate Park               Brewery         Burrito Place   
4       Harbor / Marina               Airport    Airport Food Court   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0            Restaurant                 Hotel                    Bar  
1            Steakhouse    Seafood Restaurant            Cheese Shop  
2  Caribbean Restaurant    Italian Restaurant           Climbing Gym  
3               Butcher            Restaurant             Comic Shop  
4          Airport Gate              Boutique          Boat or Ferry

<b>8. Cluster Neighbourhoods of Toronto</b>

In [44]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0], dtype=int32)

In [50]:
# add clustering labels
neighbourhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = toronto_df

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighbourhood
toronto_merged = toronto_merged.join(neighbourhoods_venues_sorted.set_index('Neighbourhood'), on='Neighbourhood')

toronto_merged.head() # check the last columns!

Postcode          Borough               Neighbourhood   Latitude  Longitude  \
0      M4N  Central Toronto               Lawrence Park  43.728020 -79.388790   
1      M4P  Central Toronto            Davisville North  43.712751 -79.390197   
2      M4R  Central Toronto          North Toronto West  43.715383 -79.405678   
3      M4S  Central Toronto                  Davisville  43.704324 -79.388790   
4      M4T  Central Toronto  Moore Park,Summerhill East  43.689574 -79.383160   

   Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0               0                  Park           Swim School   
1               0          Dance Studio        Breakfast Spot   
2               0        Clothing Store   Sporting Goods Shop   
3               0          Dessert Shop        Sandwich Place   
4               1            Playground                  Park   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0    Dim Sum Restaurant              Bus Line           Yoga Studio   
1          Burger Joint     Food & Drink Shop        Clothing Store   
2           Coffee Shop           Yoga Studio    Chinese Restaurant   
3           Pizza Place    Seafood Restaurant            Restaurant   
4                 Diner  Fast Food Restaurant        Farmers Market   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0               Dog Run  Fast Food Restaurant        Farmers Market   
1                 Hotel        Sandwich Place                  Park   
2          Dessert Shop   Rental Car Location                 Diner   
3                  Café           Coffee Shop      Sushi Restaurant   
4    Falafel Restaurant           Event Space  Ethiopian Restaurant   

  9th Most Common Venue       10th Most Common Venue  
0    Falafel Restaurant                  Event Space  
1                   Gym                  Event Space  
2    Salon / Barbershop               Sandwich Place  
3    Italian Restaurant               Farmers Market  
4     Electronics Store  Eastern European Restaurant

<b>Cluster 1</b>

In [56]:
tcluster1 = toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]
tcluster1

Borough  Cluster Labels 1st Most Common Venue  \
0    Central Toronto               0                  Park   
1    Central Toronto               0          Dance Studio   
2    Central Toronto               0        Clothing Store   
3    Central Toronto               0          Dessert Shop   
5    Central Toronto               0           Coffee Shop   
8    Central Toronto               0           Coffee Shop   
10  Downtown Toronto               0           Coffee Shop   
11  Downtown Toronto               0   Japanese Restaurant   
12  Downtown Toronto               0           Coffee Shop   
13  Downtown Toronto               0           Coffee Shop   
14  Downtown Toronto               0           Coffee Shop   
15  Downtown Toronto               0           Coffee Shop   
16  Downtown Toronto               0           Coffee Shop   
17  Downtown Toronto               0           Coffee Shop   
18  Downtown Toronto               0           Coffee Shop   
19  Downtown Toronto               0           Coffee Shop   
20  Downtown Toronto               0           Coffee Shop   
21  Downtown Toronto               0                  Café   
22  Downtown Toronto               0                   Bar   
23  Downtown Toronto               0        Airport Lounge   
24  Downtown Toronto               0           Coffee Shop   
25  Downtown Toronto               0                  Café   
26  Downtown Toronto               0                  Café   
28      East Toronto               0      Greek Restaurant   
29      East Toronto               0  Fast Food Restaurant   
30      East Toronto               0                  Café   
31      East Toronto               0    Light Rail Station   
32      West Toronto               0                Bakery   
33      West Toronto               0                   Bar   
34      West Toronto               0        Breakfast Spot   
35      West Toronto               0                  Café   
36      West Toronto               0        Breakfast Spot   
37      West Toronto               0           Coffee Shop   

   2nd Most Common Venue  3rd Most Common Venue  \
0            Swim School     Dim Sum Restaurant   
1         Breakfast Spot           Burger Joint   
2    Sporting Goods Shop            Coffee Shop   
3         Sandwich Place            Pizza Place   
5                    Pub            Pizza Place   
8         Sandwich Place                   Café   
10            Restaurant                   Park   
11      Sushi Restaurant            Coffee Shop   
12                Bakery                   Café   
13        Clothing Store                   Café   
14            Restaurant                   Café   
15            Restaurant           Cocktail Bar   
16                  Café     Italian Restaurant   
17                  Café             Steakhouse   
18                 Hotel               Aquarium   
19                 Hotel                   Café   
20                  Café                  Hotel   
21           Coffee Shop                 Bakery   
22                  Café  Vietnamese Restaurant   
23       Airport Service       Airport Terminal   
24            Restaurant                   Café   
25           Coffee Shop                  Hotel   
26         Grocery Store                   Park   
28           Coffee Shop         Ice Cream Shop   
29        Ice Cream Shop             Board Shop   
30           Coffee Shop     Italian Restaurant   
31           Yoga Studio       Recording Studio   
32              Pharmacy            Supermarket   
33           Men's Store                   Café   
34           Coffee Shop                   Café   
35    Mexican Restaurant     Italian Restaurant   
36             Gift Shop              Bookstore   
37      Sushi Restaurant                   Café   

            4th Most Common Venue      5th Most Common Venue  \
0                        Bus Line                Yoga Studio   
1               Food & Drink Shop        

<b>Cluster 2</b>

In [52]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
4   Central Toronto               1            Playground   
9  Downtown Toronto               1                  Park   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
4                  Park                 Diner  Fast Food Restaurant   
9            Playground                 Trail           Yoga Studio   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
4        Farmers Market    Falafel Restaurant           Event Space   
9        Discount Store  Fast Food Restaurant        Farmers Market   

  8th Most Common Venue 9th Most Common Venue       10th Most Common Venue  
4  Ethiopian Restaurant     Electronics Store  Eastern European Restaurant  
9    Falafel Restaurant           Event Space         Ethiopian Restaurant

<b>Cluster 3</b>

In [53]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
6  Central Toronto               2                Garden   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
6           Yoga Studio               Dog Run   Filipino Restaurant   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
6  Fast Food Restaurant        Farmers Market    Falafel Restaurant   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
6           Event Space  Ethiopian Restaurant      Electronics Store

<b>Cluster 4</b>

In [54]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
7  Central Toronto               3                  Park   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
7                 Trail      Sushi Restaurant         Jewelry Store   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
7           Yoga Studio               Dog Run  Fast Food Restaurant   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
7        Farmers Market    Falafel Restaurant            Event Space

<b>Cluster 5</b>

In [55]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
27  East Toronto               4           Coffee Shop                   Pub   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
27          Neighborhood           Yoga Studio        Discount Store   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
27  Fast Food Restaurant        Farmers Market    Falafel Restaurant   

   9th Most Common Venue 10th Most Common Venue  
27           Event Space   Ethiopian Restaurant

<b>9. Conclusion</b>

<p>There are 4 boroughs and 38 neighbourhoods in Toronto.</p>
<p>There are 1694 venues in Toronto Neighbourhood.</p>
<p>There are 235 uniques categories in venues.</p>
<p>Cluster 1 has 34 neighbourhoods with common venue categories coffe shop and restaurant/store.</p>
<p>Cluster 2 has 2 neighbourhoods with common venue categories of park and playground.</p>
<p>Cluster 3 has 1 neighbourhoods with venue categories of Garden, Yoga studio and others.</p>
<p>Cluster 4 has 1 neighbourhood  with venue category park, trail and others.</p>
<p>Cluster 5 has 1 neighbourhood   with venue category cofeeshop, pub and others.</p>